### DISCUSSION TITLE
How does everybody implement client-side validation and include the TRPC zodError from the server?

### DISCUSSION THREAD

**Discussion by Angelelz** [trpc/trpc#4656] [Q&A] [answered]
I have a monorepo setup and I share the validation schema from a package to the server and the client.
I want to do client-side validation before I send anything to the server and so far I've followed a lot of examples where you can show the zodError from the `useMutation` hook.
What I've come up with was to have another piece of state to unify the error from the server and the client as follows:
``` ts
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fieldErrors, setFieldErrors] = useState({
    username: [""],
    password: [""]
  });

  const { mutate, isLoading } = api.public.register.useMutation({
    onError(error) {
      if (error.data?.zodError?.fieldErrors) {
        const usernameErr = error.data.zodError.fieldErrors.username;
        const passwordErr = error.data.zodError.fieldErrors.password;
        setFieldErrors((state) => ({
          username: usernameErr ? usernameErr : state.username,
          password: passwordErr ? passwordErr : state.password
        }));
      }
    }
  });

  const registerHandler = () => {
    const validation = registerSchema.safeParse({ username, password });

    if (!validation.success) {
      const usernameErr = validation.error.flatten().fieldErrors.username;
      const passwordErr = validation.error.flatten().fieldErrors.password;
      console.log({ usernameErr, passwordErr });
      setFieldErrors((state) => ({
        username: usernameErr ? usernameErr : state.username,
        password: passwordErr ? passwordErr : state.password
      }));
      return;
    }

    mutate({ username, password });
  };
  return (
  ...
          <Input
            value={username}
            onChangeText={setUsername}
            placeholder="Username"
          />
          {fieldErrors.username && (
            <Text className="-mt-3 mb-2 text-red-500">
              {fieldErrors.username}
            </Text>
          )}

  )
  ...
```

Now, this works but seems a little convoluted to to have a second state.
I want to do it this way because I have additional validation in the server if the username is taken and stuff like that, where I throw zodErrors so I can show the in the client in a nice way next to the input.

Is there a better way I haven't think of yet?

**Comment by KATT** [comment, marked as answer]
This is a bit simplified version of what I do: https://kitchen-sink.trpc.io/react-hook-form

Additionally, I have `Input`-components that also ge and display field errors:

```tsx
export function useFieldError(props: { name: string }) {
  const context = useFormContext();
  if (!context) {
    throw new Error('useFieldError must be used within a Form');
  }

  const formState = useFormState({
    name: props.name,
  });

  type ErrorType = {
    message: string;
    type: string;
  };
  const error: ErrorType | undefined = get(formState.errors, props.name);

  return error;
}

export interface InputProps
  extends React.DetailedHTMLProps<
    InputHtmlAttributesWithName<HTMLInputElement>,
    HTMLInputElement
  > {
  // [...]
}

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  function Input(props, ref) {
    const fieldError = useFieldError(props);
    const error = props.ignoreErrors ? undefined : fieldError;

    // [...]
  }
)
```

**Reply by Angelelz** [reply]
I haven't ever used a form library before, mostly because I wanted to understand how does it work on my own.
I think it might be time to let a library do this work for me.
I appreciate the reply.

**Reply by Angelelz** [reply]
I dove into the example and I'm planning to clone it later today but on the surface it doesn't look like any `ZodError` from the server is being handled in the client. Please correct me if I'm wrong.
To clarify: on a register form, for example, I'm throwing a `ZodError` in the server if the username is taken, and I put it in the `fieldErrors`, so that the UI handles it just like input validation and shows the error to the user. I would like to know if there straightforward way to unify the error handling from the client validation and the server validation.
Thanks for looking into this.

**Reply by Angelelz** [reply]
You sent in the right path. Turns out that react hook form exposes a `setError` function that allows me to set the error externally.
Perfect for the `onError` function. Now I don't need to create another state.

Thanks a bunch! Great work with TRPC!
